enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,664 @@
+1. Initializing the RHEL-HA cluster
+2. Transitioning control of etcd to RHEL-HA
+3. Enabling fencing in RHEL-HA
+
+###### Transitioning etcd Management to RHEL-HA
+An important facility of the installation flow is the transition from a CEO deployed etcd to one controlled by RHEL-HA. The basic transition works as follows:
+1. [MCO extensions](https://docs.openshift.com/container-platform/4.17/machine_configuration/machine-configs-configure.html#rhcos-add-extensions_machine-configs-configure) are used to ensure that the pacemaker and corosync RPMs are installed. The installer also creates MachineConfig manifests to pre-configure resource agents.
+2. Upon detection that the cluster infrastructure is using the DualReplica controlPlaneTopology in the infrastructure config, an in-cluster entity (see open questions regarding whether this should be handled by CEO or a new 2NO setup operator) will run a command on one of the cluster nodes to initialize pacemaker. The outcome of this is that the resource agent will be started on both nodes.
+3. The aforementioned in-cluster entity will signal CEO to relinquish control of etcd by setting CEO's `managedEtcdKind` to `External`. When this happens, CEO immediately removes the etcd pod from the static pod configs. The resource agents for etcd are running from step 2, and they are configured to wait for etcd pods to be gone so they can restart them using Podman.
+4. The installation proceeds as normal once the pods start.
+If for some reason, the etcd pods cannot be started, then the installation will fail. The installer will pull logs from the control-plane nodes to provide context for this failure.
+
+There is an open question regarding how to handle updates to the etcd pod definition if it needs to change or if certificates are rotated.
+
+###### Configuring Fencing Via MCO
+Fencing setup is the last important aspect of the cluster installation. For the cluster installation to be successful, fencing should be configured and active before we declare the installation successful. To do this, baseboard management console (BMC) credentials need to be made available to the control-plane nodes as part of pacemaker initialization.
+To ensure rapid fencing using pacemaker, we will collect RedFish details (address, username, and **password**) for each node via the install-config (see proposed install-config changes).
+This will take a format similar to that of the [Baremetal Operator](https://docs.openshift.com/container-platform/4.17/installing/installing_bare_metal_ipi/ipi-install-installation-workflow.html#bmc-addressing_ipi-install-installation-workflow).
+We will create a new MachineConfig that writes BMC credentials to the control-plane disks. This will resemble the BMC specification used by the [BareMetalHost](https://docs.openshift.com/container-platform/4.17/rest_api/provisioning_apis/baremetalhost-metal3-io-v1alpha1.html#spec-bmc) CRD.
```

Comment:
The latter. BMO stores the URL and a link to a credentials secret in a BareMetalHost CRD.
For our own implementation, it is simpler to load these on disk, since we don't have anything to consume these in-cluster.

Reply:
Are there any security concerns tied to adding this new credentials file to disk? Does this increase our expose?

Reply:
I think we want to rely on k8s's Secret management here rather than roll a new thing. MachineConfigs themselves are not encrypted at rest AFAIK.
Can the process that needs the credentials be deployed as a DaemonSet with a Secret volume mounted into it?

Reply:
I'll look into the specifics of using a daemonset. The concern I have is in the case the cluster enters a state where etcd quorum is lost, is that volume still mounted to disk? I think the answer is yes, since that shouldn't affect the scheduling of existing containers.

I think this would work for the initial setup. It would probably also be possible to handle secret updates with this, by allowing the user to update the secrets file when the cluster is healthy.

The biggest concern is when the configuration becomes invalid when the cluster is unhealthy. In this case, you need to run a reinitialization process directly on the nodes to update the secret. This may be fine, but I want to review with the SMEs to ensure that the situations where this happens and where a full recovery would be impossible without manual intervention are exactly 1 to 1.

Reply:
Running workloads can keep running without an API server. Although if the Pod is dead at the same time as the API server goes down you are in trouble.

A static Pod would be better but you [can't mount Secrets in them](https://kubernetes.io/docs/concepts/configuration/secret/#restriction-static-pod) :slightly_frowning_face: